### PR TITLE
[BEAM-10650] OrderedListState API

### PR DESCRIPTION
**Pull request by reuvenlax**
This PR adds the API and and in-memory implementation for the timestamp-ordered list state. The API is currently marked experimental and is still subject to change.

### CODE REVIEW

**Comment by reuvenlax** [comment]
Run Portable_Python PreCommit

**Comment by reuvenlax** [comment]
R: @dpmills

File: sdks/java/core/src/main/java/org/apache/beam/sdk/state/OrderedListState.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+ * can be added to the list, and subranges can be read out in order. Subranges of the list can be
+ * deleted as well.
+ */
+@Experimental(Kind.STATE)
+public interface OrderedListState<T>
+    extends GroupingState<TimestampedValue<T>, Iterable<TimestampedValue<T>>> {
+
+  /** Read the entire list ordered by timestamp. */
+  @Override
+  Iterable<TimestampedValue<T>> read();
```
Comment (dpmills) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/state/OrderedListState.java:36]: Don't need to include the inherited methods here.  If you do include them, include all of them, including isEmpty

File: sdks/java/core/src/main/java/org/apache/beam/sdk/state/OrderedListState.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+  /**
+   * Read a timestamp-limited subrange of the list. The result is ordered by timestamp.
+   *
+   * <p>All values with timestamps >= minTimestamp and < limitTimestamp will be in the resuling
```
Comment (dpmills) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/state/OrderedListState.java:41]: Using an exclusive upper limit means that Instant(Long.MAX_VALUE) isn't a supported timestamp.  Do we care?  Regardless, document the exact supported timestamps for this interface

File: sdks/java/core/src/main/java/org/apache/beam/sdk/state/OrderedListState.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  /**
+   * Clear a timestamp-limited subrange of the list.
+   *
+   * <p>All values with timestamps >= minTimestamp and < limitTimestamp will be removed from the
+   * list.
+   */
+  void clearRange(Instant minTimestamp, Instant limitTimestamp);
+
+  @Override
+  OrderedListState<T> readLater();
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/state/OrderedListState.java]: Done

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillStateInternals.java [hunk 4]
```diff
@@ -456,6 +470,64 @@ protected WorkItemCommitRequest persistDirectly(WindmillStateCache.ForKey cache)
...
+      this.namespace = namespace;
+      this.spec = spec;
+      this.stateKey = encodeKey(namespace, spec);
+      this.stateFamily = stateFamily;
+      this.elemCoder = elemCoder;
+    }
+
+    @Override
+    public Iterable<TimestampedValue<T>> read() {
+      return null;
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillStateInternals.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/state/OrderedListState.java (hunk 1 above)
Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/state/OrderedListState.java]: Removed

File: sdks/java/core/src/main/java/org/apache/beam/sdk/state/OrderedListState.java (hunk 2 above)
Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/state/OrderedListState.java:36]: Added documentation to this effect.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/state/OrderedListState.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  /**
+   * Clear a timestamp-limited subrange of the list.
+   *
+   * <p>All values with timestamps >= minTimestamp and < limitTimestamp will be removed from the
+   * list.
+   */
+  void clearRange(Instant minTimestamp, Instant limitTimestamp);
+
+  OrderedListState<T> readRangeLater(Instant minTimestamp, Instant limitTimestamp);
```
Comment (dpmills) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/state/OrderedListState.java:50]: Add a method comment

**Comment by reuvenlax** [comment]
run dataflow validatesrunner

File: sdks/java/core/src/main/java/org/apache/beam/sdk/state/OrderedListState.java (hunk 5 above)
Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/state/OrderedListState.java:54]: done

**Comment by reuvenlax** [comment]
run flink validatesrunner

**Comment by reuvenlax** [comment]
retest this please

**Comment by reuvenlax** [comment]
Run Dataflow ValidatesRunner

**Comment by reuvenlax** [comment]
retest this please

**Comment by damgad** [comment]
retest this please

**Comment by reuvenlax** [comment]
Run Dataflow ValidatesRunner

**Comment by reuvenlax** [comment]
run flink validatesrunner

**Comment by reuvenlax** [comment]
run spark validatesrunner

**Comment by reuvenlax** [comment]
run flink portable validatesrunner

**Comment by reuvenlax** [comment]
run flink validatesrunner

**Comment by reuvenlax** [comment]
run flink portable validatesrunner

**Comment by reuvenlax** [comment]
run spark validatesrunner

**Comment by reuvenlax** [comment]
Run Dataflow ValidatesRunner

**Comment by reuvenlax** [comment]
run flink validatesrunner

**Comment by reuvenlax** [comment]
run spark validatesrunner

**Comment by reuvenlax** [comment]
run flink portable validatesrunner

**Comment by reuvenlax** [comment]
Run Dataflow ValidatesRunner

**Comment by reuvenlax** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by reuvenlax** [comment]
@kennknowles I am merging based on @dpmills LGTM. If you see any more things that should be changed, please comment and I will open a new PR to address those comments.

**Pull request merged**